pkg/controller/clusterapimachine/machine_controller.go

```diff
@@ -0,0 +1,301 @@
+			return
+		}
+		m, ok = tombstone.Obj.(*capiv1.Machine)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a Machine %#v", obj))
+			return
+		}
+	}
+	clustoplog.WithClusterAPIMachine(c.logger, m).Debugf("deleting machine")
+	c.enqueueMachine(m)
+}
+
+// Run runs c; will not return until stopCh is closed. workers determines how
+// many machines will be handled in parallel.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Infof("starting %s controller", controllerLogName)
+	defer c.logger.Infof("shutting down %s controller", controllerLogName)
+
+	if !controller.WaitForCacheSync("capi-machine", stopCh, c.machinesSynced) {
+		c.logger.Info("waiting for caches to sync")
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(ms *capiv1.Machine) {
+	key, err := controller.KeyFunc(ms)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", ms, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Sync %q failed with %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncMachine will sync the machine with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncMachine(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing machine")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machine")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	machine, err := c.machinesLister.Machines(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("machine has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	mLog := clustoplog.WithClusterAPIMachine(c.logger, machine)
+	mLog.Debugf("found machine")
+
+	// Lookup the cluster for this machine:
+	clusterName, ok := machine.Labels[clusterNameLabel]
+	if !ok {
+		return fmt.Errorf("no cluster label set on machine")
+	}
+	cluster, err := c.clustersLister.Clusters(namespace).Get(clusterName)
+	if errors.IsNotFound(err) {
+		mLog.WithField("cluster", clusterName).Errorf("no cluster found for machine")
+		return nil
```

Comment:
A Machine may be created prior to the Cluster with which the Machine is associated. In that case, the controller will not process the Machine until the Machine is resynced from the API Server. We should be watching Clusters and reqeueueing Machines when on updates of their Clusters.

Reply:
I assume this is in addition to requeueing on Machine changes.

Would it be reasonable to only requeue on Cluster creation (not update), as I don't think there's any action to take on a cluster update right now, at this point I still think this is just a snapshot in time once it's set.

Reply:
No, I don't think that is reasonable.
1) The user should be able to update the Cluster.
2) We will very soon be inspecting the ClusterVersionRef from the Status, which will certainly be done on an update rather than at creation.

Reply:
(2) has me convinced. Updated with update/delete cluster listeners.

Reply:
Why are you not swayed by (1)? What do you expect the behavior of cluster-operator to be when a user changes the spec of a cluster?